Rate limiter

edobrb opened the pull request:
Inspired by [Cloudflare](https://blog.cloudflare.com/counting-things-a-lot-of-different-things/)
Closes #40 

API example:
```typescript
import { functions } from '@mondrian-framework/module'
import { rateLimitMiddleware } from '@mondrian-framework/rate-limiter'
import { RedisSlotProvider, SlotProvider } from '@mondrian-framework/rate-limiter'
import { createClient } from '@redis/client'

const redisClient = process.env.REDIS_URL ? createClient() : undefined
redisClient?.on('error', (err) => console.log('Redis Client Error', err))
redisClient?.connect()
export const slotProvider: SlotProvider | undefined = redisClient && new RedisSlotProvider(redisClient)


const loginRateLimit = rateLimitMiddleware<typeof loginData, typeof user, typeof loginError, LoginContext>({
  key: ({ input }) => input.email,
  rate: '100 requests in 1 hour',
  onLimit: () => Promise.resolve(result.fail({ tooManyRequests: 'Too many requests. Retry in few minutes.' })),
  slotProvider
})

export const login = functions.withContext<LoginContext>().build({
  input: loginData,
  output: user,
  error: loginError,
  body: async ({ input, context }) => {
    // ...
  },
  middlewares: [loginRateLimit],
})
```

edobrb committed:
feat: closes #40

edobrb committed:
chore: format

giacomocavalieri commented on packages/middleware/rate-limiter/src/rate.ts:
```diff
@@ -0,0 +1,33 @@
+
+/**
+ * {@link Rate} expressed by a string
+ */
```
This documentation needs at least a couple of examples

giacomocavalieri commented on packages/middleware/rate-limiter/src/rate.ts:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Rate description.
+ */
+export type Rate = {
```
This documentation is not enough imo. I'd need it to answer a couple questions:
- What is a rate
- How does it relate to rate limiting
- What are some examples of rates

edobrb committed:
chore: add rate-limiter README

giacomocavalieri commented on packages/middleware/rate-limiter/src/slot.ts:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Slot provider.
+ */
```
Why put a docstring if it only has the name of the interface? It should answer some questions:
- what is a slot provider?
- what is its purpose?
- how does it relate to rate limiting?
- what are examples of different slot providers and how do they affect the rate limiting process?

giacomocavalieri commented on packages/middleware/rate-limiter/src/slot.ts:
```diff
@@ -0,0 +1,27 @@
+
+/**
+ * Sliding window slot.
+ */
```
Same as above, I have no idea what a sliding window slot and the comment isn't helping me understand the _purpose_ of this interface

giacomocavalieri commented on packages/middleware/rate-limiter/src/slot.ts:
```diff
@@ -0,0 +1,27 @@
+   * Slot starting time.
+   */
+  readonly fromSecond: number
+  /**
+   * Increases slot internlal counter.
+   */
+  inc(): void
+  /**
+   * Gets slot internal counter.
+   */
+  value(): number
```
- what does the "internal counter" mean?
- what does it represent?
- is it just an ever-increasing integer?
- does this pose a risk of overflowing?
- how should these methods be used? And what would I use these methods for?

edobrb replied:
This is a critical point, need a good documentation because if someone want to implement another support (in-memory, redis, ...) need to understand this

edobrb replied:
No risk of overflowing in my opinion. No risk of performing 9007199254740991 requests without never restarting the server or going to the next slot.

giacomocavalieri commented on packages/middleware/rate-limiter/src/implementation/in-memory.ts:
```diff
@@ -0,0 +1,25 @@
+  create({ fromSecond }: { fromSecond: number }): Slot {
+    return new InMemorySlot(fromSecond)
+  }
+}
```
Add documentation that explains:
- What it means to be an `InMemorySlot`
- When should one use this Provider and not other ones

giacomocavalieri commented on packages/middleware/rate-limiter/src/implementation/redis.ts:
```diff
@@ -0,0 +1,78 @@
+    durationSecond: number
+    key: string
+    client: RedisClientType
+  }) {
```
This big annotation makes the code really hard to read, it should be a type on its own and not inlined here

giacomocavalieri commented on packages/middleware/rate-limiter/src/implementation/redis.ts:
```diff
@@ -0,0 +1,78 @@
+    this.counter = 0
+    this.fromSecond = fromSecond
+    this.durationSecond = durationSecond
+    this.key = `${key}:${fromSecond}`
```
Why is the key built like this? What invariants should hold? Is it expected to be unique? I'd add a comment documenting this aspect

giacomocavalieri commented on packages/middleware/rate-limiter/src/slot.ts:
```diff
@@ -0,0 +1,27 @@
+  /**
+   * Slot starting time.
+   */
+  readonly fromSecond: number
```
Why call this `fromSecond` if it is the slot's starting time, it's really confusing. I'd change the name to `startingTime`

edobrb replied:
Agree, but i always want the unit if the type does not old it.

```
const startingTimeSeconds = 0
//or
type Seconds = number
const startingTime: Seconds = 0
```

giacomocavalieri replied:
Calling it `startingTypeSeconds` or using the type alias is perfect for me, already makes the intent clearer! 👍

By the way, it is not worth it in typescript since there's no opaque types, but this is a place where you could use a phantom type in other languages. If you want an interesting read about it I can recommend you [this blog post](https://blog.hayleigh.dev/phantom-types-in-gleam) :)

giacomocavalieri commented on packages/middleware/rate-limiter/src/slot.ts:
```diff
@@ -0,0 +1,27 @@
+  /**
+   * Creates an active slot for the given period.
+   */
+  create(args: { fromSecond: number; durationSecond: number; key: string }): Slot
```
This doesn't explain what each arg is for:
- what is `fromSecond` (by the way, I think it should be called `startingTime`)
- what is `durationSecond`, what does it do? How does it affect the generated slot?
- what is `key`, what is that used for?

giacomocavalieri commented on packages/middleware/rate-limiter/src/implementation/redis.ts:
```diff
@@ -0,0 +1,78 @@
+      .then((value) => {
+        if (this.counter === 0) {
+          this.client
+            .expireAt(this.key, new Date((this.fromSecond + this.durationSecond + 60) * 1000), 'NX')
```
I don't get what the resulting date is going to be, can you explain it to me?

edobrb replied:
The expiration is set to the end time of the slot (fromSecond + durationSecond) + 1 minute (60) in order to avoid premature deletion if different servers are not perfectly time aligned. I'll add a const for the 60.

giacomocavalieri commented on packages/middleware/rate-limiter/src/implementation/redis.ts:
```diff
@@ -0,0 +1,78 @@
+  }
+
+  value(): number {
+    this.client
+      .get(this.key)
+      .catch(() => console.error('REDIS: Failed to get', this.key))
+      .then((value) => {
+        if (value != null) {
+          const v = Number(value)
+          if (!Number.isNaN) {
+            this.counter = v
+          }
+        }
+      })
```
What is this promise doing? Also does this action need to be over _before_ we return `this.counter`? It looks strange to me that a call to `value()` might return a value and then it gets changed _after_ behind the scened by the promise you started here

edobrb replied:
No need to be over. We don't want to slow down the request only to be 100% sure the user can call again the service. One or two requests that should be rate-limited but still pass is not a big problem.

giacomocavalieri replied:
Thanks for taking the time to explain this to me!
Maybe you could add it to a comment explaining this, to avoid that future maintainers will be puzzled by this since it's not immediately clear from the code

giacomocavalieri commented on packages/middleware/rate-limiter/src/implementation/redis.ts:
```diff
@@ -0,0 +1,78 @@
+      .then((value) => {
+        if (value != null) {
+          const v = Number(value)
+          if (!Number.isNaN) {
```
This is most definitely a typo, did you mean to write `!Number.isNaN(v)`?

giacomocavalieri commented on packages/middleware/rate-limiter/src/implementation/redis.ts:
```diff
@@ -0,0 +1,78 @@
+
+/**
+ * Redis slot provider. Uses a redis client to provide rate limiting capabilities for any middleware.
+ * It can be used when a service scale horizzontally.
```
Did you mean to write?
```suggestion
 * It can be used when a service needs to scale horizontally.
```

giacomocavalieri commented on packages/middleware/rate-limiter/src/implementation/redis.ts:
```diff
@@ -0,0 +1,78 @@
+  constructor(client: RedisClientType<any, any, any>, keyPrefix: string = 'mondrian-rate-limiter:') {
+    this.client = client
+    this.keyPrefix = keyPrefix
+  }
```
Missing documentation that explains what each argument is used for

giacomocavalieri commented on packages/middleware/rate-limiter/src/implementation/redis.ts:
```diff
@@ -0,0 +1,78 @@
+
+  create(args: { fromSecond: number; durationSecond: number; key: string }): Slot {
+    return new RedisSlot({ ...args, key: `${this.keyPrefix}${args.key}`, client: this.client })
+  }
```
Missing documentation that explains what each argument is used for and how they influence the slot creation

giacomocavalieri commented on packages/middleware/rate-limiter/src/middleware.ts:
```diff
@@ -0,0 +1,64 @@
+
+/**
+ * Limits a function execution with the given rate limit.
+ * You can group different calls toghether by using the key function. If the key function returns `null` not rate limit is applied.
```
```suggestion
 * You can group different calls toghether by using the key function. If the key function returns `null` no rate limiting is applied.
```

giacomocavalieri commented on packages/middleware/rate-limiter/src/middleware.ts:
```diff
@@ -0,0 +1,64 @@
+    slotProvider: options.slotProvider ?? new InMemorySlotProvider(),
+  })
+  return {
+    name: 'Rate limiter - Leacky bucket',
```
```suggestion
    name: 'Rate limiter - Leaky bucket',
```

giacomocavalieri commented on packages/middleware/rate-limiter/src/middleware.ts:
```diff
@@ -0,0 +1,64 @@
+  options: {
+    rate: Rate | RateLiteral
+    slotProvider?: SlotProvider
+  }
```
This type annotation makes the code really hard to read, it should be a type on its own and not inlined like this imo

giacomocavalieri commented on packages/middleware/rate-limiter/src/middleware.ts:
````diff
@@ -0,0 +1,64 @@
+ *   onLimit: () => Promise.resolve(result.fail({ tooManyRequests: 'Too many requests. Retry in few minutes.' })),
+ * })
+ * ```
+ */
````
This documentation doesn't explain how each argument is going to be used:
- What are the options used for?
- When is `onLimit` going to be called? What happens with its return type? What happens if it returns an error?
- Why is the options.provider optional? What if it's undefined? What is its default value? Is the default good for most use cases?
- Why is `onLimit` optional, what happens if its undefined?

giacomocavalieri commented on packages/middleware/rate-limiter/src/middleware.ts:
```diff
@@ -0,0 +1,64 @@
+      if (onLimit) {
+        return onLimit(args)
+      }
+      throw new Error(`Too many requests`)
```
I'd rewrite this using `else if`s, in general it makes it easier to follow the execution flow
```suggestion
      if (res === 'allowed') {
        return next(args)
      } else if (onLimit) {
        return onLimit(args)
      } else {
        throw new Error(`Too many requests`)
      }
```

giacomocavalieri commented on packages/middleware/rate-limiter/src/rate.ts:
```diff
@@ -0,0 +1,33 @@
+  /**
+   * Period unit.
+   */
+  unit: Unit
```
Is this doc really helpful?
![image](https://github.com/twinlogix/mondrian-framework/assets/20598369/547a4165-4960-43a1-9305-190076e2f57b)

Jokes aside, I'd remove this if it's just repeating the name of the field. What would be best, in my opinion, is providing a longer description of what these fields are and how they define a slot. For example, for `period`:

```
/*
 * This is the time period spanned by the Rate. It is an absolute number and
 * its unit is defined by the `unit` field.
 *
 * ## Examples
 * 
 * Consider for example the following rates:
 * 
 * ```ts
 * const rate1 = parseRate("10 requests in 10 seconds")
 * const rate2 = parseRate("20 requests in 10 minutes")
 * ```
 * 
 * Both the rates have the same `period` of 10, but different units so
 * their rate limiting effects will be vastly different.
 */
```

giacomocavalieri commented on packages/middleware/rate-limiter/src/rate.ts:
```diff
@@ -0,0 +1,33 @@
+/**
+ * {@link Rate} expressed by a string
+ */
+export type RateLiteral = `${number} requests in ${number} ${Unit}`
```
The only gripe I have with this definition is that its going to have a grammatical error if the number of messages/minutes/seconds is 1:

```
const rate1 = "1 requests in 1 minutes"
const rate2 = "1 requests in 1 seconds"
```

giacomocavalieri commented on packages/middleware/rate-limiter/src/implementation/redis.ts:
```diff
@@ -0,0 +1,78 @@
+      .get(this.key)
+      .catch(() => console.error('REDIS: Failed to get', this.key))
+      .then((value) => {
+        if (value != null) {
```
Why are you using `!=` instead of `!==`, what are the possible values of `value`?

giacomocavalieri commented on packages/middleware/rate-limiter/src/rate.ts:
```diff
@@ -0,0 +1,33 @@
+export type Unit = 'seconds' | 'minutes' | 'hours'
```
Missing documentation, it should at least explain what this is used for

giacomocavalieri commented on packages/middleware/rate-limiter/src/sliding-window.ts:
```diff
@@ -0,0 +1,125 @@
+  private readonly rate: Rate
+  private readonly slotProvider: SlotProvider
+
+  constructor({ rate, slotProvider }: { rate: Rate | RateLiteral; slotProvider: SlotProvider }) {
```
This is missing documentation that explains how the arguments are used to build a provider and what each argument does

giacomocavalieri replied:
Also I think it should only accept a `Rate`, it's not this class responsibility to parse a `RateLiteral` into a `Rate`; it should be the outside caller that has to provide it an already valid `Rate`

giacomocavalieri commented on packages/middleware/rate-limiter/src/sliding-window.ts:
```diff
@@ -0,0 +1,125 @@
+    this.slotProvider = slotProvider
+  }
+
+  get({ key }: { key: string }): SlidingWindow {
```
Why is it taking a single field object and not just a string argument? What's the reasoning behind this?

giacomocavalieri commented on packages/middleware/rate-limiter/src/sliding-window.ts:
```diff
@@ -0,0 +1,125 @@
+
+  get({ key }: { key: string }): SlidingWindow {
+    const slidingWindow = this.slidingWindows.get(key)
+    if (slidingWindow) {
+      return slidingWindow
+    }
+    const newSlidingWindow = new SlidingWindow({ rate: this.rate, slotProvider: this.slotProvider, key })
+    this.slidingWindows.set(key, newSlidingWindow)
+    return newSlidingWindow
```
This is a matter of personal preference but I'd use the explicit `else` branch, it also makes it easier to see the flow of the code and refactor things in the future

```suggestion
    if (slidingWindow) {
      return slidingWindow
    } else {
      const newSlidingWindow = new SlidingWindow({ rate: this.rate, slotProvider: this.slotProvider, key })
      this.slidingWindows.set(key, newSlidingWindow)
      return newSlidingWindow
    }
```

giacomocavalieri commented on packages/middleware/rate-limiter/src/sliding-window.ts:
```diff
@@ -0,0 +1,125 @@
+  }
+}
+
+export class SlidingWindow {
```
Missing documentation for exported class:
- what is a sliding window?
- where is it used and what is it used for?
- does it even make sense for someone to make a `new SlidingWindow()`? Or should the provider the only one allow to create new ones?

giacomocavalieri commented on packages/middleware/rate-limiter/src/sliding-window.ts:
```diff
@@ -0,0 +1,125 @@
+  private readonly key: string
+  private rateLimitedUntil?: number
+
+  constructor({
+    rate,
+    nowSeconds,
+    slotProvider,
+    key,
+  }: {
+    rate: Rate
+    nowSeconds?: () => number
+    slotProvider: SlotProvider
+    key: string
+  }) {
```
The inlined type annotation makes the code harder to read, I'd move it into its own type

Also it would be useful to have documentation that explains what each argument does:
- what is the rate used for?
- what is `nowSeconds`? Why is it a function? What is it used for? What happens if it's undefined? What happens if the function throws? Does it really need to be a function?
- what is key? Why does it need it?

giacomocavalieri commented on packages/middleware/rate-limiter/src/sliding-window.ts:
```diff
@@ -0,0 +1,125 @@
+    if (rate.requests < 0) {
+      throw new Error('Rate limit must be a positive duration')
+    }
+    this.samplingPeriodSeconds = rate.period * (rate.unit === 'hours' ? 3600 : rate.unit === 'minutes' ? 60 : 1)
```
This violates the single responsibility principle, it's not the sliding windows' responsibility to make this conversion. It should be the `Rate` (that is the object holding all the data) the one able to returns a `second` duration

giacomocavalieri commented on packages/middleware/rate-limiter/src/sliding-window.ts:
```diff
@@ -0,0 +1,125 @@
+
+  private getSlots(): [Slot, Slot] {
+    const now = this.nowSeconds()
+    const actualFrom = Math.floor(now / this.samplingPeriodSeconds) * this.samplingPeriodSeconds
```
What are these division and multiplication doing?

edobrb replied:
I'll try my best to document all the math

giacomocavalieri commented on packages/middleware/rate-limiter/src/sliding-window.ts:
```diff
@@ -0,0 +1,125 @@
+    const actualFrom = Math.floor(now / this.samplingPeriodSeconds) * this.samplingPeriodSeconds
+    const oldFrom = actualFrom - this.samplingPeriodSeconds
+
+    const oldSlot =
+      this.slots.get(oldFrom) ??
+      this.slotProvider.create({
+        fromSecond: oldFrom,
+        durationSecond: this.samplingPeriodSeconds,
+        key: this.key,
+      })
+    this.slots.set(oldFrom, oldSlot)
+    const actualSlot =
+      this.slots.get(actualFrom) ??
+      this.slotProvider.create({
+        fromSecond: actualFrom,
+        durationSecond: this.samplingPeriodSeconds,
+        key: this.key,
+      })
+    this.slots.set(actualFrom, actualSlot)
```
Could you explain to me, in natural language, what's the purpose of this block of code?

edobrb replied:
Get the two slot (old and actual)
<img width="797" alt="image" src="https://github.com/twinlogix/mondrian-framework/assets/50401517/90f5213c-2b29-4127-b72c-3f2968d7d485">
If they does not exists -> creates them.

After create a slot save it to the map `slots` so it can be retrieved later.

giacomocavalieri replied:
Maybe it could be clearer to add another internal function `getOrCreate` to factor out these two pieces of code since they're kind of similar? Also it would be better at self-documenting its purpose

giacomocavalieri commented on packages/middleware/rate-limiter/src/sliding-window.ts:
```diff
@@ -0,0 +1,125 @@
+      })
+    this.slots.set(actualFrom, actualSlot)
+
+    //Remove expired slots reference
+    if (this.slots.size > 2) {
+      const slots = [...this.slots.values()].sort((slot1, slot2) => slot2.fromSecond - slot1.fromSecond)
+      for (let i = 2; i < slots.length; i++) {
+        this.slots.delete(slots[i].fromSecond)
+      }
+    }
```
Why not make this an internal function named `removeExpiredSlots`? It would document itself better and make the `getSlots` function smaller and easier to read

giacomocavalieri commented on packages/middleware/rate-limiter/src/sliding-window.ts:
```diff
@@ -0,0 +1,125 @@
+
+  inc(): 'allowed' | 'rate-limited' {
+    const now = this.nowSeconds()
+    if (this.rateLimitedUntil != null && now <= this.rateLimitedUntil) {
```
Why are you using `!=` can `this.rateLimitedUntil` be `undefined` as well?

edobrb replied:
yes != null check if !== null and !== undefined

giacomocavalieri commented on packages/middleware/rate-limiter/src/sliding-window.ts:
```diff
@@ -0,0 +1,125 @@
+    this.key = key
+  }
+
+  private getSlots(): [Slot, Slot] {
```
`getSlots` is quite a cryptic name, it doesn't really explain anything except the fact that you're returning some slots.
- Why are exactly two slots returned?
- Do these slots have any particular property?
- What are the invariants holding for these slots?

Also this function is not simply returning slots but is also updating the object's internal state by removing references to other slots. I wouldn't expect a simple "getter" function to have such behaviour, its name should reflect that:
`getSlotsXYZAndRemoveOldReferences`.

If you think it looks bad and is too long then the function may be split in two different functions, but anyways I prefer a longer name rather than letting unexpected behaviour slip in my code by accident

giacomocavalieri commented on packages/middleware/rate-limiter/src/sliding-window.ts:
```diff
@@ -0,0 +1,125 @@
+    //Remove expired slots reference
+    if (this.slots.size > 2) {
+      const slots = [...this.slots.values()].sort((slot1, slot2) => slot2.fromSecond - slot1.fromSecond)
+      for (let i = 2; i < slots.length; i++) {
```
Why is this excluding the first two slots? Are those guaranteed to never be expired? If so, why? What piece of code guarantees that? There should be a comment explaining this

giacomocavalieri commented on packages/middleware/rate-limiter/src/sliding-window.ts:
```diff
@@ -0,0 +1,125 @@
+    const oldValue = oldSlot.value()
+    const currentValue = currentSlot.value()
+    const oldPart = (this.samplingPeriodSeconds - (now - currentSlot.fromSecond)) / this.samplingPeriodSeconds
+    const requests = currentValue + oldPart * oldValue
```
What is requests? Could you explain to me why you used this formula?

giacomocavalieri commented on packages/middleware/rate-limiter/src/sliding-window.ts:
```diff
@@ -0,0 +1,125 @@
+        : ((currentValue - this.rateLimit) * this.samplingPeriodSeconds) / oldValue +
+          this.samplingPeriodSeconds +
+          currentSlot.fromSecond
+    return 'rate-limited'
```
I'd rather have an explicit else than having to stop for a second to see what happens to the control flow after the if
```suggestion
    } else {
      this.rateLimitedUntil =
        oldValue === 0
          ? currentSlot.fromSecond + this.samplingPeriodSeconds
          : ((currentValue - this.rateLimit) * this.samplingPeriodSeconds) / oldValue +
            this.samplingPeriodSeconds +
            currentSlot.fromSecond
      return 'rate-limited'
    }
```

giacomocavalieri commented on packages/middleware/rate-limiter/src/sliding-window.ts:
```diff
@@ -0,0 +1,125 @@
+        ? currentSlot.fromSecond + this.samplingPeriodSeconds
+        : ((currentValue - this.rateLimit) * this.samplingPeriodSeconds) / oldValue +
+          this.samplingPeriodSeconds +
+          currentSlot.fromSecond
```
Can you explain me what this formula does and how?

minox86 commented:
@giacomocavalieri just a comment on your comments: try to invest the time better by suggesting changes or additions, including in the documentation, not only comments. ;-)

edobrb commented:
![image](https://github.com/twinlogix/mondrian-framework/assets/50401517/9766fc30-f82f-4227-8349-e80441f33ece)

I didn't think a draft pull request deserved so much attention, however I'll try to change/responde all your reviews :D

giacomocavalieri commented:
Sorry for the many questions, but I cannot add documentation to things I don't understand, that's why I pointed out to Edo the points where documentation would be needed, otherwise he's the only one who can read and fix this code :)

As for the little fixes I used actionable fixes so that he can apply those if he thinks those are ok and not risk any conflict

giacomocavalieri commented:
> ![image](https://user-images.githubusercontent.com/50401517/271307165-9766fc30-f82f-4227-8349-e80441f33ece.png)
> 
> I didn't think a draft pull request deserved so much attention, however I'll try to change/responde all your reviews :D

You requested a review from me, that's why I left it 😆

edobrb committed:
Update packages/middleware/rate-limiter/src/implementation/redis.ts

Co-authored-by: Giacomo Cavalieri <giacomo.cavalieri@icloud.com>

edobrb committed:
Update packages/middleware/rate-limiter/src/middleware.ts

Co-authored-by: Giacomo Cavalieri <giacomo.cavalieri@icloud.com>

edobrb committed:
chore: fixes for review

edobrb committed:
Update packages/middleware/rate-limiter/src/middleware.ts

Co-authored-by: Giacomo Cavalieri <giacomo.cavalieri@icloud.com>

edobrb committed:
Update packages/middleware/rate-limiter/src/sliding-window.ts

Co-authored-by: Giacomo Cavalieri <giacomo.cavalieri@icloud.com>

edobrb committed:
chore: fixes for review

edobrb committed:
Merge branch 'feature/rate-limit' of https://github.com/twinlogix/mondrian-framework into feature/rate-limit

edobrb committed:
fix: minor for review

edobrb committed:
fix: minor for review

edobrb committed:
doc: add documentation to rate-limiter

edobrb committed:
fix: minor bug in redis

edobrb committed:
chore: add redis test in rate-limiter

edobrb committed:
fix: minor

edobrb commented:
I've added documentation at all critical points. In my opinion it is more than enough, if some points are unclear please report them. I also refactored some small parts to make them more readable.

The only doubt I have left is about the interface exposed to the user -> that of the middleware. @giacomocavalieri and I had talked about and for the moment we couldn't think of another way.

edobrb committed:
chore: minor refactor

edobrb committed:
chore: update README
